### ISSUE TITLE
About Scene Reconstruction Performance

### ISSUE THREAD

**Issue by Luciferbobo** [astra-vision/SceneRF#7]
Hi all. I'm trying to evaluate this method on Semantic KITTI using the provided pretrained model. But I got some less-than-one numbers in all three metrics. Orz
![image](https://user-images.githubusercontent.com/54936491/212048336-a3b7004d-c5cd-4e97-a823-3d1c1bc623f6.png)

I checked the TSDF results and found that it seemed like the prediction of voxels was strange. Below is a comparison of ground truth (left) and my prediction (right).

![3@FHZFK793~PYPH~ M%NI~O](https://user-images.githubusercontent.com/54936491/212050030-b2021bdd-edd0-4021-9307-da51aef0e569.png)

I was wondering if you had any insight about what might be happening. Thx!

**Comment by anhquancao** [comment]
Hi, which TSDF threshold did you use? 
Also, when drawing, you need to remove the voxels with label 255.

**Comment by Luciferbobo** [comment]
Truncated margin was set to 10. I used default settings in depth2tsdf.py and fusion.py. Thanks for your kindly reminding. I wanted to see the overall result so I put all the labels together :)

**Comment by anhquancao** [comment]
The 255 denotes unknown voxels which are voxels that have no lidar rays pass through in all the sequence. 
Do you use this function to generate the occupancy?
https://github.com/astra-vision/SceneRF/blob/main/scenerf/scripts/evaluation/eval_sr.py#L11
The idea is to set an increase threshold since the depth error increase with the distance to the vehicle
Because if you consider all voxels within +-10m around the depth as occupied then it means around 20/0.2=100 voxels are set to be occupied for each depth value.

**Comment by Luciferbobo** [comment]
Yes. I used this function and the default settings in eval_sr.py. I tried to ignore the 255 label and I saw the corresponding no lidar area in GT disappeared, but my prediction result still looks the same.

**Comment by anhquancao** [comment]
Is it the same for other frame? Did you try to draw the mesh? Do the rendered depth images look fine?

**Comment by Luciferbobo** [comment]
I drew some frames and they all looked the same... Corresponding depth and RGB image looks fine. I attached them below.

![image](https://user-images.githubusercontent.com/54936491/212106605-b90775e5-b155-4b29-9f21-db2009d30bed.png)
![image](https://user-images.githubusercontent.com/54936491/212106691-7a1b8233-7b79-4aba-bb08-6c7e34fd50bf.png)

**Comment by anhquancao** [comment]
Thanks, let me check it!
What if you change the variable th here to a smaller number? is it still the same?
https://github.com/astra-vision/SceneRF/blob/main/scenerf/scripts/evaluation/eval_sr.py#L12

**Comment by anhquancao** [comment]
It should look like this. 
![Screenshot from 2022-11-02 14-42-49](https://user-images.githubusercontent.com/9668858/212110463-37ef12c8-ed63-4910-8273-177eb5068109.png)
This what I drew months ago. I will need to check

**Comment by anhquancao** [comment]
You can obtain the mesh from this line
https://github.com/astra-vision/SceneRF/blob/main/scenerf/scripts/reconstruction/depth2tsdf.py#L107

**Comment by Luciferbobo** [comment]
I tried default th=0.25 and th=0.15. They still look the same :(
![image](https://user-images.githubusercontent.com/54936491/212112323-b6593c5e-1fea-48d0-a06e-d12a2e7467fd.png)
Many thanks for your prompt response! Very great work. I will check it also.

**Comment by anhquancao** [comment]
Sorry for this, probably I changed smth when cleaning it. 
I will update it in few weeks.

**Comment by anhquancao** [comment]
Seem like the threshold is not applied at all.

**Comment by Luciferbobo** [comment]
OK. I will look forward to.

**Comment by Luciferbobo** [comment]
OK. I will look forward to it. w

**Comment by anhquancao** [comment]
Great! You can watch the repo, so you will be notified when I update it.

**Comment by anhquancao** [comment]
Hi @Luciferbobo,
Sorry, I am swamped, I just tried to draw the occupancy prediction using the following code:

```
def get_grid_coords(dims, resolution):
    '''
    :param dims: the dimensions of the grid [x, y, z] (i.e. [256, 256, 32])
    :return coords_grid: is the center coords of voxels in the grid
    '''

    # The sensor in centered in X (we go to dims/2 + 1 for the histogramdd)
    g_xx = np.arange(0, dims[0] + 1)
    # The sensor is in Y=0 (we go to dims + 1 for the histogramdd)
    g_yy = np.arange(0, dims[1] + 1)
    # The sensor is in Z=1.73. I observed that the ground was to voxel levels above the grid bottom, so Z pose is at 10
    # if bottom voxel is 0. If we want the sensor to be at (0, 0, 0), then the bottom in z is -10, top is 22
    # (we go to 22 + 1 for the histogramdd)
    # ATTENTION.. Is 11 for old grids.. 10 for new grids (v1.1) (https://github.com/PRBonn/semantic-kitti-api/issues/49)
    sensor_pose = 10
    g_zz = np.arange(0, dims[2] + 1)

    # Obtaining the grid with coords...
    xx, yy, zz = np.meshgrid(g_xx[:-1], g_yy[:-1], g_zz[:-1])

    coords_grid = np.array([xx.flatten(), yy.flatten(), zz.flatten()]).T
    coords_grid = coords_grid.astype(np.float)

    coords_grid = (coords_grid * resolution) + resolution/2

    temp = np.copy(coords_grid)
    temp[:, 0] = coords_grid[:, 1]
    temp[:, 1] = coords_grid[:, 0]
    coords_grid = np.copy(temp)

    return coords_grid, g_xx, g_yy, g_zz


def draw(
    voxels,
    cam_param_path="",
    voxel_size=0.04):    

    voxels[voxels == 255] = 0
    grid_coords, _, _, _ = get_grid_coords([voxels.shape[0], voxels.shape[1], voxels.shape[2]], voxel_size)    

    points = np.vstack([grid_coords.T, voxels.reshape(-1)]).T

    # Obtaining voxels with semantic class
    points = points[(points[:, 3] != 0)]
    
    vis = o3d.visualization.Visualizer()
    vis.create_window(width=1200, height=600)
    ctr = vis.get_view_control()
    param = o3d.io.read_pinhole_camera_parameters(cam_param_path)

    pcd = o3d.geometry.PointCloud()
    pcd.points = o3d.utility.Vector3dVector(points[:, :3])
    pcd.estimate_normals()
    vis.add_geometry(pcd)

    ctr.convert_from_pinhole_camera_parameters(param)

    vis.run()  # user changes the view and press "q" to terminate
    param = vis.get_view_control().convert_to_pinhole_camera_parameters()
    o3d.io.write_pinhole_camera_parameters(cam_param_path, param)
    
path = "Your path to stored TSDF output"
frame_id = "001385.npy"
tsdf_path = os.path.join(path, frame_id)
tsdf = np.load(tsdf_path)

occ = np.zeros_like(tsdf)
occ[tsdf > 0.2 ] = 0
occ[abs(tsdf) < 0.2 ] = 1
draw(occ)
```
This is the output:
![image](https://user-images.githubusercontent.com/9668858/223999868-ff53cf12-e738-49c1-a413-008c0fcf5396.png)

**Comment by Luciferbobo** [comment]
Hi. Apologize for also being swamped with other things lately. Thank you very much for the update! I tried the function you shared, while the results still seem to have some issues... Here's a comparison between the GT (left) and my prediction (right). 
![image](https://user-images.githubusercontent.com/54936491/226306150-c984c724-6db4-461b-9bd0-9e68e1a7f9dd.png)
![image](https://user-images.githubusercontent.com/54936491/226306182-2ccbcafd-59a6-4183-91bc-06409e7b9936.png)

I guess the visualization results produced by Open3D and Matplotlib Axes3D should be similar, so the issue might be with the code in depth2tsdf.py. I must admit that I'm not very familiar with the parameter settings in the TSDFVolume function. I was wondering whether the visualization issue may be due to the parameter settings in this section of the code? Thx! =v=

**Comment by anhquancao** [comment]
Hi @Luciferbobo,
Thank you for your information! I just found a bug related to the reconstruction and have updated them. Could you please try cloning again?

**Comment by Luciferbobo** [comment]
The reconstructed results from the updated code is looking much better, and I can clearly see a more reasonable FOV. I believe some minor details may still need improvement due to insufficient training epochs. I will try evaluating all reconstruction metrics once I finished my training. Thx for the update! 
![image](https://user-images.githubusercontent.com/54936491/226515271-4f6510c9-2f00-4556-a716-52d7aa962750.png)

**Comment by anhquancao** [comment]
Awesome! I'm so glad it's working for you. Let me know if there's anything else I can do to help. 😊
